### DISCUSSION TITLE
[cc65] Wrong behavior for long int and compute management

### DISCUSSION THREAD

**Discussion by jedeoric** [cc65/cc65#2034] [Q&A]
Long is wrong when we compute a unsigned char and unsigned long. It produces a signed long

    /*******************************/
    #include <stdio.h>
    void main()
    {
    // UNsigned
     unsigned char kernel_header[2];
     // UNsigned
     unsigned long ul;
     kernel_header[1]=0xbc;
     kernel_header[0]=0x07;
     // ul is declared with unsigned long and kernel_header as unsigned char
     ul = (kernel_header[1])*256 + kernel_header[0];
     printf("ul: 0x%lx\n", ul);
     }
    /*******************************/

=> ul: 0xFFFFBC07

Workaround to fix this with last version of cc65 :

    /*******************************/
    #include <stdio.h>
    void main()
    {
    // UNsigned
    unsigned char kernel_header[2];

    // UNsigned
    unsigned long ul;

    kernel_header[1]=0xbc;
    kernel_header[0]=0x07;

    // Cast UNsigned int
    ul = (unsigned int) ((kernel_header[1])*256) + kernel_header[0];

    printf("ul: 0x%lx\n", ul);
    }
    /*******************************/
=> ul: 0xBC07

The behavior is strange because we declare both *unsigned* for kernel_header and ul

**Comment by mrdudz** [comment]
doesnt it happen because kernel_header[1]*256 is implicitly converted to int?

**Comment by kugelfuhr** [comment]
The behavior is correct. 

Both unsigned chars on the right are converted to ints (integer promotion). So the right side evaluates to a (signed!) integer.

For the assignment to unsigned long, ISO/IEC 9899:1990 (E) (the C90 standard) says in 6.2.1.2 "Signed and unsigned integers":

"When a signed integer is converted to an unsigned integer with equal or greater size, if the value of the signed integer is nonnegative, its value is unchanged. Otherwise: if the unsigned integer has greater size, the signed integer is first promoted to the signed integer corresponding to the unsigned integer: the value is converted to unsigned by adding to it one greater than the largest number that can be represented in the unsigned integer type ”

Footnote 28 is attached to this paragraph which reads:

"In a two‘s-complement representation. there is no actual change in the bit pattern except filling the high-order bits with copies of the sign bit if the unsigned integer has greater size"

So the int is first promoted to a long int, and then assigned to the unsigned long without any changes of the internal representation.

gcc does it the same way. The program

```C
#include <stdio.h>
int main()
{
    int x = 0xbc070000;
    unsigned long y = x;
    printf("%08lX\n", y);
    return 0;
}
```

outputs "FFFFFFFFBC070000" on operating systems where "unsigned long" is larger than "int".

**Comment by mrdudz** [comment]
I couldnt have explained it like this, thanks :)

**Reply by jedeoric** [reply]
Thanks for the explanation. It's clear.
